### ISSUE TITLE
Calling `borrow_mut` on a `Box`ed trait object and passing the result to a function can cause a spurious compile error

### ISSUE THREAD

**Issue by SolraBizna** [rust-lang/rust#50872]
This code does not compile:

```rust
use std::borrow::BorrowMut;

pub trait Trait {}
pub struct Struct {}
impl Trait for Struct {}

fn func(_: &mut Trait) {}

fn main() {
    let mut foo: Box<Trait> = Box::new(Struct{});
    func(foo.borrow_mut());
}
```

However, if you use `as` to be clear on what you expect `borrow_mut()` to return:

```rust
    func(foo.borrow_mut() as &mut Trait);
```

It will then compile.

This does not happen when the boxed type is a struct, only a trait. Simple assignments don't seem to trigger it either.

Here is the output when compiling on the Playground:

```
   Compiling playground v0.0.1 (file:///playground)
error[E0597]: `foo` does not live long enough
  --> src/main.rs:11:10
   |
11 |     func(foo.borrow_mut());
   |          ^^^ borrowed value does not live long enough
12 | }
   | - `foo` dropped here while still borrowed
   |
   = note: values in a scope are dropped in the opposite order they are created

warning: variable does not need to be mutable
  --> src/main.rs:10:9
   |
10 |     let mut foo: Box<Trait> = Box::new(Struct{});
   |         ----^^^
   |         |
   |         help: remove this `mut`
   |
   = note: #[warn(unused_mut)] on by default

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by cuviper** [comment]
It also works if you deref the box, `func((*foo).borrow_mut())`, or even just `func(&mut(*foo))`.

**Comment by Spoonbender** [comment]
Triage: no change
